### ISSUE TITLE
Python 3: update string literals to have either u, r, or b prefix

### ISSUE THREAD

**Issue by feerrenrut** [nvaccess/nvda#9637]
### Problem:

In python 3, plain string literals become unicode strings. In some parts of NVDA we use plain strings for binary data. This will cause errors and strange behaviour.

To gain confidence in a Python 3 release of NVDA we need consider these strings, and how they are used, and decide if the literal should be prefixed with 'u', 'r', or 'b'.
There are more than 1000 unspecified string literals in the NVDA codebase. To count this I used regex `[^u\n"]"[^"\n]` on `*.py` files in the `repo/source` directory, pycharm IDE also allows to ignore results in comments.

### Suggested approach:
 - Existing cases specifying 'u' are ok, they were already intended to be used as unicode strings.
 - Existing cases specifying 'r' are higher risk, they may be used for binary data. We should look at these first.
    - Using regex `r".+"` there seems to be just under 5000 of these strings.
 - Cases with no prefix, the vast majority will be ok to be unicode strings. There are certainly some cases that are used as bytes / binary data. These will be the hardest to find.

Looking at each string individually will take weeks, I suggest we see how we can exclude low risk areas:

- translated strings ( `_("blah")`, `pgettext("blah")`) can be ignored. Whether these have 'u', 'r', or no prefix. We can be quite confident they will not be bytes.

Open questions:
- How are we going to keep track of what has been looked at / excluded?
  - Is it feasible to use regex to automate adding string literals to areas that we are confident in?
- How do we identify general cases we can exclude?

**Comment by feerrenrut** [comment]
Part of difficulty here is crafting reliable regex to find these cases.

**Comment by feerrenrut** [comment]
One option here might be to implement a fixer in 2to3 to help find the cases we care about. 2to3 already contains a parser, and the appropriate framework for running it over a large set of files. I will look at this approach first. I'll see if I can get it to add a a unique comment to the source that we can then search for.

**Comment by leonardder** [comment]
Note that we do not only have to deal with the b literal for byte 
strings, but in that case, indexing behavior is also changed and we need 
to account for that.

**Comment by leonardder** [comment]
>     * Existing cases specifying 'r' are higher risk, they may be used for binary data. We should look at these first.
>       
>       * Using regex `r".+"` there seems to be just under 5000 of these strings.

Have you considered cases like `r'.+'` here? I guess this regex also catches strings like `r"""hello"""`, in which case we should also match on line breaks.

>     * Cases with no prefix, the vast majority will be ok to be unicode strings. There are certainly some cases that are used as bytes / binary data. These will be the hardest to find.

Most of them will probably be in braille display drivers. We could make hwIo treat unicode strings in braille display drivers as latin-1, in which case the bytes mapping is the same. However, that really obfuscates possible issues.

>     * How are we going to keep track of what has been looked at / excluded?

May be we should process the source code package by package or area by area. For example, brailleDisplayDrivers, braille, bdDetect and hwIo are heavily intertwined. These are also the risky ones. Another risky case might be textInfos, though NVDA is pretty demanding on enforcing unicode on these strings already, so that's good.

>       * Is it feasible to use regex to automate adding string literals to areas that we are confident in?

As soon as we still double check, yes. It doesn't mean that we have to look at every string, but we need t oread through all the code and hopefully catch issues if they are there.
 
>     * How do we identify general cases we can exclude?

What do you mean with that?

**Comment by feerrenrut** [comment]
@leonardder I think there are a few misunderstandings here. Rather than answer your questions individually I'll explain my progress on this so far.

My initial plan was to look at every string. I was expecting to use regex to find the next string missing a 'u' prefix, asses it, and modify it to have a 'u' prefix. This allows us to keep track of what as been assessed, and is comprehensive. I thought first I would see what this looks like. The first regex I used to just to get some indication of the number of strings that did not use 'u'. This had over ten thousand matches. Given 30 seconds to look at each, this will take more than 80 hours. So probably not a great approach, this is a long time and will be error prone, also I expect most of the strings we look at are fine. I have also realised that is hard or perhaps impossible to write reliable regex for this, essentially a parser is needed.

To be clear, lets restate the goal of this:
**We want to locate strings that may have been treated as binary data in NVDA on python 2.7, and determine if the code must be modified to the same behaviour with python 3.7.**

What do these strings look like:
- 'r' strings
  - These were previously raw strings, not unicode.
  - These either stay as 'r' unicode strings, or become 'b' strings.
- Strings with no prefix.
  - Previously not unicode

**Other requirements for the approach**
- We want to exclude strings we have already looked at.
   - This might take a several days, and we may have to go over the code base multiple times.
   - we want to be sure we are done.
- In final merge, don't change lines that don't need to be changed.
   - For instance, we may keep track of lines by prefixing a 'u' character, or with a comment.
   - If there is no functional change, it would be best if these changes were not included.

Looking at some of the regex matches I got, I realised that we can categorise the matches if we consider more of the context on the line. For instance, strings passed to the translation system are almost certainly not going to cause an issue once converted to unicode. This is one general case we can exclude, we don't have to look at these strings to determine if they are safe as unicode. Another case is 'r' strings that are passed to `re` eg `re.compile(r'something')`, these almost certainly don't require attention, and they account for a large proportion of the 'r' strings. We may be able to identify other general cases that are also safe.

Currently I am investigating whether we can locate these strings using something smarter than regex. Ideally this would automate:
- Finding the strings
- Filtering out categories of usage that we know are safe (passed to translation or regex)
- Tag those that need attention with an easy to search for comment.

We would then be able to look at those results, and see if there are any other categories we can also exclude. Then we have a system we can re-run to confirm we have comprehensively checked the codebase.

At the moment I am looking at modifying 2to3 to do this. It [seems quite extensible](http://python3porting.com/fixers.html).

Of course, another approach is "just guess" which files are more likely to contain problems. Looking at those specifically. This is the fall back plan if we can't do something more comprehensive.
